feat: add logic for eager evaluation in REPL

Vinit-Pandit opened the pull request:
Progress: #2073.

## Description

> What is the purpose of this pull request?

This pull request:

-   add draft logic for eager evaluation feature in `REPL`

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   Progress: #2073.

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

Vinit-Pandit committed:
feat: add draft logic of eager evaluation

Vinit-Pandit committed:
fix: fixing test failure

Vinit-Pandit commented on lib/node_modules/@stdlib/repl/lib/main.js:
```diff
@@ -363,6 +381,7 @@ function REPL( options ) {
 		FLG = self._editorActions.beforeKeypress( data, key );
 
+		self._eagerEvaluator.beforeKeypress( data, key );
 		// If ENTER keypress is encountered or if a preview was completed while navigating, gracefully close the completer...
```
```suggestion

		// If ENTER keypress is encountered or if a preview was completed while navigating, gracefully close the completer...
```

Vinit-Pandit replied:
```suggestion

		// If ENTER keypress is encountered or if a preview was completed while navigating, gracefully close the completer...
```

Vinit-Pandit committed:
fix: fixing test failure

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

Vinit-Pandit committed:
checking options.setting

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: passed
---

Vinit-Pandit committed:
fix: style fix

Signed-off-by: Vinit Pandit <106718914+MeastroZI@users.noreply.github.com>

Vinit-Pandit commented:
@kgryte Please take a look at the logic when you have time. If everything looks good, I can proceed with:
1) Fixing conflicts with the Tab preview
1) Adding tests and additional files for eager evaluation"

aayush0325 committed:
style: fix indentation in manifest.json files

PR-URL: https://github.com/stdlib-js/stdlib/pull/4602

Co-authored-by: aayush0325 <ayush25khanna@gmail.com>
Reviewed-by: Philipp Burckhardt <pburckhardt@outlook.com>

Vinit-Pandit commented:
![img6](https://github.com/user-attachments/assets/ce09dcea-b710-4912-bf7d-22ef642172f5)
![img5](https://github.com/user-attachments/assets/877ec8a2-6c07-428e-8094-0e2072e5527f)
![img4](https://github.com/user-attachments/assets/be85e26e-b614-4512-8fcd-db235910ab10)
![img3](https://github.com/user-attachments/assets/6a23b3a4-6df1-486c-baa5-efb7d97da5c5)
![img2](https://github.com/user-attachments/assets/ca4a1f36-8244-4270-8b1c-1d38a28605e0)
![img1](https://github.com/user-attachments/assets/667c3d0d-40d7-4168-95d0-ec9c611ad330)


here are some Images

Vinit-Pandit commented:
/stdlib rebase

Vinit-Pandit committed:
Merge branch 'stdlib-js:develop' into eager_evaluation_repl

Vinit-Pandit commented:
/stdlib update-copyright-years

stdlib-bot committed:
chore: update copyright years

Vinit-Pandit commented:
I request to have one look on the [logic](https://github.com/MeastroZI/stdlibMine/edit/eager_evaluation_repl/lib/node_modules/%40stdlib/repl/lib/eager_evaluation.js?pr=%2Fstdlib-js%2Fstdlib%2Fpull%2F4277)

Snehil-Shah commented on lib/node_modules/@stdlib/repl/lib/defaults.js:
```diff
@@ -106,7 +106,10 @@ function defaults() {
-			'theme': 'stdlib-ansi-basic'
+			'theme': 'stdlib-ansi-basic',
+
+			// Flag indicating whether to enable eager Evaluation (note: default depends on whether TTY):
```
```suggestion
			// Flag indicating whether to enable eager evaluation (note: default depends on whether TTY):
```

Snehil-Shah commented on lib/node_modules/@stdlib/repl/lib/eager_evaluation.js:
```diff
@@ -0,0 +1,267 @@
+
+// VARIABLES //
+
+var debug = logger( 'repl:EE' );
```
the logger name can be more verbose

Snehil-Shah commented on lib/node_modules/@stdlib/repl/lib/eager_evaluation.js:
```diff
@@ -0,0 +1,267 @@
+// MAIN //
+
+/**
+* Constructor for creating a multi-line handler.
```
update description

Snehil-Shah commented on lib/node_modules/@stdlib/repl/lib/eager_evaluation.js:
```diff
@@ -0,0 +1,267 @@
+* @param {Boolean} enabled - boolean indicating whether the syntax-highlighter should be initially enabled
+* @returns {EagerEvaluator} eager-evaluator instance
+*/
+function EagerEvaluator( repl, rli, enabled ) {
+	this._repl = repl;
+	this._rli = rli;
+	this._cmd = repl._cmd;
+	this._enabled = enabled;
+	this._isPreview = false;
+
+	return this;
+}
```
I would suggest adding comments describing each variable for the sake of consistency (see other handlers)

Snehil-Shah commented on lib/node_modules/@stdlib/repl/lib/eager_evaluation.js:
```diff
@@ -0,0 +1,267 @@
+	var res;
+	var tmp;
+
+	if (!this._enabled || this._rli.line === '' || key.name === 'paste-start' || (key.name === 'o' && key.ctrl) || key.name === 'return') {
+		return;
+	}
```
I'm pretty sure this check is an overkill.. any chance it can be avoided? Many handlers automatically clear the screen below before their operations, so you might not actually have to worry about all these triggers..

Snehil-Shah commented on lib/node_modules/@stdlib/repl/lib/eager_evaluation.js:
```diff
@@ -0,0 +1,267 @@
+* @returns {void}
+*
+*/
+setNonEnumerableReadOnly(EagerEvaluator.prototype, 'onKeypress', function onKeyPress(data, key) {
```
```suggestion
setNonEnumerableReadOnly(EagerEvaluator.prototype, 'onKeypress', function onKeypress(data, key) {
```

Snehil-Shah commented on lib/node_modules/@stdlib/repl/lib/eager_evaluation.js:
```diff
@@ -0,0 +1,267 @@
+	};
+	if (!this.isSideEffectFree(code)) {
+		debug('code have side effect');
+		return;
+	}
+	debug('try to process command');
+	tmp = processCommand(code);
+	if (tmp instanceof Error) {
+		debug( 'getting error %s', tmp.message );
+		return;
```
Here and everywhere else, add spaces inside parentheses for readability..

Snehil-Shah commented on lib/node_modules/@stdlib/repl/lib/eager_evaluation.js:
```diff
@@ -0,0 +1,267 @@
+* @returns {void}
+*
+*/
+setNonEnumerableReadOnly(EagerEvaluator.prototype, 'beforeKeypress', function beforeKeyPress(data, key) {
```
```suggestion
setNonEnumerableReadOnly(EagerEvaluator.prototype, 'beforeKeypress', function beforeKeypress(data, key) {
```

Snehil-Shah commented on lib/node_modules/@stdlib/repl/lib/eager_evaluation.js:
```diff
@@ -0,0 +1,267 @@
+});
+
+/**
+* Tells weather code is side effect free or not.
```
typos

Snehil-Shah commented on lib/node_modules/@stdlib/repl/lib/eager_evaluation.js:
```diff
@@ -0,0 +1,267 @@
+	* Function which recursivly traverse from the node and tells weather node is side effect free or not.
+	*
+	* @private
+	* @param {JSON} node - ast node.
```
```suggestion
	* @param {Object} node - ast node
```
here and everywhere

Snehil-Shah commented on lib/node_modules/@stdlib/repl/lib/eager_evaluation.js:
```diff
@@ -0,0 +1,267 @@
+	*
+	* @private
+	* @param {JSON} node - ast node.
+	* @returns {Boolean} - Boolean indicate weather node is side effect free or not.
```
```suggestion
	* @returns {Boolean} - boolean indicating whether the node is side effect free or not
```

Snehil-Shah commented on lib/node_modules/@stdlib/repl/lib/main.js:
```diff
@@ -160,6 +161,14 @@ function REPL( options ) {
 	opts.settings.autoDisableBracketedPasteOnExit = ( opts.settings.autoDisableBracketedPasteOnExit === void 0 ) ? opts.isTTY : opts.settings.autoDisableBracketedPasteOnExit; // eslint-disable-line max-len
 	opts.settings.syntaxHighlighting = ( opts.settings.syntaxHighlighting === void 0 ) ? opts.isTTY : opts.settings.syntaxHighlighting; // eslint-disable-line max-len
 
+	// This is temparary because eager evaluation conflicting with the some tests.
+	if ( options && options.settings && options.settings.eagerEvaluation !== void 0 ) {
+		opts.settings.eagerEvaluation = options.settings.eagerEvaluation;
+	}
```
Pinning as a reminder

Snehil-Shah replied:
@Vinit-Pandit Okay, just noticed that you haven't yet implemented a setting yet to toggle eager evaluation (that's what causing the clash in the tests). Here are some references that'll help you make the necessary changes: https://github.com/stdlib-js/stdlib/blob/develop/lib/node_modules/%40stdlib/repl/lib/main.js#L1605, https://github.com/stdlib-js/stdlib/blob/develop/lib/node_modules/%40stdlib/repl/lib/settings.js

Snehil-Shah commented on lib/node_modules/@stdlib/repl/lib/main.js:
```diff
@@ -254,6 +263,12 @@ function REPL( options ) {
 	// Initialize an internal flag indicating whether we've received a `SIGINT` signal:
 	setNonEnumerable( this, '_SIGINT', false );
 
+	// Initialize an internal flag indicating whether command is eagerlyEvaluated:
+	setNonEnumerable( this, '_isEagerEvaluated', false );
+
+	// Initialize as internal variable for caching the compiled object from eagerEvaluation
+	setNonEnumerable( this, '_eagerEvaluatedExecutable', void 0 );
+
```
I get the performance benefits of caching the compiled object that was used during eager evaluation, but IMO we can afford to avoid this. Also because, the upcoming REPL refactoring will abstract away the compilation and execution logic, and entagling it here will just make it difficult. Basically, we can afford re-compiling the command during execution instead of relying on cache from the eager evaluator, so I would suggest keeping those 2 processes separate.

Snehil-Shah commented on lib/node_modules/@stdlib/repl/lib/main.js:
```diff
@@ -363,6 +381,8 @@ function REPL( options ) {
 		completed = self._previewCompleter.beforeKeypress( data, key );
 		FLG = self._editorActions.beforeKeypress( data, key );
 
```
```suggestion
```

Snehil-Shah reviewed (changes requested):
PS: I was NOT able to reproduce the behavior shown in the GIFs, hence couldn't test it out myself. Something needs fixing..

Vinit-Pandit commented:
Got it i will look into it , thank you for your time.

Vinit-Pandit committed:
fix: fixing clashing in end line of terminal

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

Vinit-Pandit committed:
fix: removing unnessary conditions check

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

Vinit-Pandit committed:
fix: adding suggested changes

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

Vinit-Pandit committed:
style: adding spaces in parenthesis

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

Vinit-Pandit commented:
i think this failures are because of some bug in github workflow as i am seeing this in every prs.

Snehil-Shah committed:
refactor: change file name for consistency

Signed-off-by: Snehil Shah <snehilshah.989@gmail.com>

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

Signed-off-by: Snehil Shah <snehilshah.989@gmail.com>

Snehil-Shah committed:
fix: add guard against invalid class initialization

Signed-off-by: Snehil Shah <snehilshah.989@gmail.com>

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

Snehil-Shah committed:
refactor: use existing constant for hardcoded values

Signed-off-by: Snehil Shah <snehilshah.989@gmail.com>

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

Snehil-Shah committed:
refactor: move entangled logic outside and clean ups

Signed-off-by: Snehil Shah <snehilshah.989@gmail.com>

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

Snehil-Shah committed:
fix: add stricter timeout and clean ups

Signed-off-by: Snehil Shah <snehilshah.989@gmail.com>

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

Snehil-Shah committed:
fix: update incorrect logic for handling multiline inputs

Signed-off-by: Snehil Shah <snehilshah.989@gmail.com>

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

Snehil-Shah commented:
@Vinit-Pandit Okay, it's looking great. Thanks for working on this! The only things left are: resolving [this](https://github.com/stdlib-js/stdlib/pull/4277#discussion_r1934410778) comment, and the internal database of pure functions, then it's good to go.

cc @kgryte

Snehil-Shah committed:
style: update debug logs

Signed-off-by: Snehil Shah <snehilshah.989@gmail.com>

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

Vinit-Pandit committed:
feat: add logic to setting to toggle eagerevaluator

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

Vinit-Pandit commented:
Thanks, @Snehil-Shah, for your time! I have implemented the logic for toggling eager evaluation.

Snehil-Shah committed:
fix: make timeout even stricter

Signed-off-by: Snehil Shah <snehilshah.989@gmail.com>

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

Snehil-Shah committed:
style: clean ups

Signed-off-by: Snehil Shah <snehilshah.989@gmail.com>

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

kgryte commented on lib/node_modules/@stdlib/repl/lib/eager_evaluator.js:
```diff
@@ -0,0 +1,330 @@
+	'ecmaVersion': 'latest'
+};
+var ROPTS = {
+	'timeout': 100, // (in milliseconds) shouldn't be much noticeable when naturally typing
```
```suggestion
	'timeout': 100, // (in milliseconds) this controls how long eagerly evaluated commands have to execute; we need to avoid setting this too high in order to avoid eager evaluation interfering with the UX when naturally typing
```

kgryte commented on lib/node_modules/@stdlib/repl/lib/eager_evaluator.js:
```diff
@@ -0,0 +1,330 @@
+	this._rli = rli;
+
+	// Cache a reference to the command array:
+	this._cmd = repl._cmd;
```
Note for @Snehil-Shah: in a future refactoring, we should make the command array a read-only class instance for consumers (e.g., internal functions) which only need read-only access. Should there be a module which explicitly needs to manipulate the command array, that could be through a separate parent class which allows mutation. In general, passing around and caching an array like this makes me nervous, as it makes debugging more difficult due to not knowing who or what is potentially mutating the command array.

Snehil-Shah replied:
Makes sense. Although, the contents of the `cmd` array is pretty short lived, and is supposed to be frequently mutated such as when a new line is added or previous lines are edited in a multiline command, and is cleaned up after every execution as well.

kgryte commented on lib/node_modules/@stdlib/repl/lib/eager_evaluator.js:
```diff
@@ -0,0 +1,330 @@
+
+	try {
+		ast = parse( code, AOPTS );
+	}
+	catch ( err ) {
```
```suggestion
	} catch ( err ) {
```

kgryte commented on lib/node_modules/@stdlib/repl/lib/eager_evaluator.js:
```diff
@@ -0,0 +1,330 @@
+	* @param {Object} node - ast node
+	* @returns {Boolean} - boolean indicating whether the node is side-effect-free
+	*/
+	function traverse( node ) {
```
Can this function be moved to the parent scope? I don't believe that the function's implementation depends on anything from the `isSideEffectFree` context.

kgryte commented on lib/node_modules/@stdlib/repl/lib/eager_evaluator.js:
```diff
@@ -0,0 +1,330 @@
+			if ( traverse( node.left ) && traverse( node.right ) ) {
+				return true;
+			}
+		}
+		else if ( node.type === 'ExpressionStatement' ) {
+			if ( traverse( node.expression ) ) {
+				return true;
+			}
+		}
+		else if ( node.type === 'CallExpression' ) {
```
```suggestion
		} else if ( node.type === 'ExpressionStatement' ) {
			if ( traverse( node.expression ) ) {
				return true;
			}
		} else if ( node.type === 'CallExpression' ) {
```

In general, we avoid placing `else if` on a separate line unless we need to add comments before the `else` clause.

kgryte commented on lib/node_modules/@stdlib/repl/lib/eager_evaluator.js:
```diff
@@ -0,0 +1,330 @@
+		else if ( node.type === 'CallExpression' ) {
+			fname = getFunctionName( node.callee );
+			if ( tempDB[fname] && tempDB[fname].isPure ) {
+				// Iterating through arguments:
```
```suggestion
				// Examine each function argument for potential side-effects...
```

kgryte commented on lib/node_modules/@stdlib/repl/lib/eager_evaluator.js:
```diff
@@ -0,0 +1,330 @@
+	* @param {Object} node - ast node
+	* @returns {string} - function name representing the node
+	*/
+	function getFunctionName( node ) {
```
Can this function be moved to parent scope?

kgryte commented on lib/node_modules/@stdlib/repl/lib/eager_evaluator.js:
```diff
@@ -0,0 +1,330 @@
+			return '';
+		}
+		if ( node.type === 'MemberExpression' ) {
+			return getFunctionName(node.object) + '_' + node.property.name;
```
```suggestion
			return getFunctionName( node.object ) + '_' + node.property.name;
```

kgryte commented on lib/node_modules/@stdlib/repl/lib/eager_evaluator.js:
```diff
@@ -0,0 +1,330 @@
+		}
+		return '';
+	}
+} );
```
```suggestion
});
```

kgryte commented on lib/node_modules/@stdlib/repl/lib/eager_evaluator.js:
```diff
@@ -0,0 +1,330 @@
+* @memberof EagerEvaluator.prototype
+* @type {Function}
+* @returns {void}
+*
```
```suggestion
```

kgryte commented on lib/node_modules/@stdlib/repl/lib/eager_evaluator.js:
```diff
@@ -0,0 +1,330 @@
+} );
+
+/**
+* Disable the eager evaluator.
```
```suggestion
* Disables the eager evaluator.
```

kgryte commented on lib/node_modules/@stdlib/repl/lib/eager_evaluator.js:
```diff
@@ -0,0 +1,330 @@
+* @param {string} data - input data
+* @param {(Object|void)} key - key object
+* @returns {void}
+*
```
```suggestion
```

kgryte commented on lib/node_modules/@stdlib/repl/lib/eager_evaluator.js:
```diff
@@ -0,0 +1,330 @@
+* @param {string} data - input data
+* @param {(Object|void)} key - key object
+* @returns {void}
+*
```
```suggestion
```

kgryte commented on lib/node_modules/@stdlib/repl/lib/eager_evaluator.js:
```diff
@@ -0,0 +1,330 @@
+	}
+
+	// Build the final command:
+	cmd = deepCopy( this._cmd );
```
I believe `this._cmd` is a plain array, correct? If so, use `array/base/copy` or `array/base/slice` instead for better performance. Only use `deepCopy` is you need to actually deep copy.

kgryte commented on lib/node_modules/@stdlib/repl/lib/eager_evaluator.js:
```diff
@@ -0,0 +1,330 @@
+	}
+
+	res = inspect( res );
+	nlInd = res.indexOf( '\n' );
```
`nlInd` is an awkward variable name. Perhaps use `index` or `ind` instead?

kgryte commented on lib/node_modules/@stdlib/repl/lib/eager_evaluator.js:
```diff
@@ -0,0 +1,330 @@
+	readline.moveCursor( this._repl._ostream, 0, -1 );
+	readline.cursorTo( this._repl._ostream, cursorPosition + this._repl.promptLength() );
+	this._isPreviewing = true;
+	debug( 'sucess' );
```
```suggestion
	debug( 'Successfully evaluated command.' );
```

kgryte commented on lib/node_modules/@stdlib/repl/lib/main.js:
```diff
@@ -293,6 +296,9 @@ function REPL( options ) {
 	// Initialize a syntax-highlighter:
 	setNonEnumerableReadOnly( this, '_syntaxHighlighter', new SyntaxHighlighter( this, this._ostream, this._settings.syntaxHighlighting ) );
 
+	// Initialize a eagerEvaluator:
```
```suggestion
	// Initialize an eager evaluator:
```

kgryte reviewed (changes requested):
Overall, this is looking good! Thanks for working on this, @Vinit-Pandit and @Snehil-Shah!

kgryte commented:
For the function database, I think we can address that in a follow-up PR. It's likely that we'll want to indicate whether APIs are pure at the package level.

Vinit-Pandit committed:
style: clean ups

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

Vinit-Pandit committed:
style: clean up

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

Vinit-Pandit committed:
style: clean up

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

Snehil-Shah committed:
style: correctly place functions

Signed-off-by: Snehil Shah <snehilshah.989@gmail.com>

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

kgryte commented on lib/node_modules/@stdlib/repl/lib/eager_evaluator.js:
```diff
@@ -0,0 +1,327 @@
+*
+* @private
+* @param {Object} node - ast node
+* @returns {Boolean} - boolean indicating whether the node is side-effect-free
```
```suggestion
* @returns {boolean} boolean indicating whether the node is side-effect-free
```

kgryte committed:
docs: fix return type

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/repl/lib/eager_evaluator.js:
```diff
@@ -0,0 +1,327 @@
+// FUNCTIONS //
+
+/**
+* Recursively traverses the node to determine whether the node is side-effect-free or not.
```
```suggestion
* Recursively traverses the node to determine whether the node is side-effect-free.
```

kgryte committed:
docs: update description

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/repl/lib/eager_evaluator.js:
```diff
@@ -0,0 +1,327 @@
+*
+* @private
+* @param {Object} node - ast node
+* @returns {string} - function name representing the node
```
```suggestion
* @returns {string} function name representing the node
```

kgryte committed:
docs: remove hyphen

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/repl/lib/eager_evaluator.js:
```diff
@@ -0,0 +1,327 @@
+}
+
+/**
+* Gets a function name representing identifier and member expression nodes.
```
```suggestion
* Resolves the function name associated with a provided AST node.
```

kgryte committed:
docs: update description

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/repl/lib/eager_evaluator.js:
```diff
@@ -0,0 +1,327 @@
+/* eslint-disable max-len */
```
```suggestion
```

kgryte commented on lib/node_modules/@stdlib/repl/lib/eager_evaluator.js:
```diff
@@ -0,0 +1,327 @@
+* limitations under the License.
+*/
+
+/* eslint-disable no-underscore-dangle, no-restricted-syntax, no-invalid-this */
```
```suggestion
/* eslint-disable no-underscore-dangle, no-restricted-syntax, no-invalid-this, max-len */
```

kgryte committed:
Apply suggestions from code review

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/repl/lib/eager_evaluator.js:
```diff
@@ -0,0 +1,326 @@
+* @private
+* @param {REPL} repl - repl instance
+* @param {Object} rli - readline instance
+* @param {Boolean} enabled - boolean indicating whether the eager evaluator should be initially enabled
```
```suggestion
* @param {boolean} enabled - boolean indicating whether the eager evaluator should be initially enabled
```

`Boolean` indicates a class instance. That is not what is meant here. `boolean` means a boolean primitive.

kgryte committed:
docs: fix type

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/repl/lib/eager_evaluator.js:
```diff
@@ -0,0 +1,326 @@
+* @type {Function}
+* @param {string} code - input code
+* @returns {boolean} - boolean indicating whether the code is side-effect-free
+*
```
```suggestion
```

kgryte commented on lib/node_modules/@stdlib/repl/lib/eager_evaluator.js:
```diff
@@ -0,0 +1,326 @@
+* @memberof EagerEvaluator.prototype
+* @type {Function}
+* @param {string} code - input code
+* @returns {boolean} - boolean indicating whether the code is side-effect-free
```
```suggestion
* @returns {boolean} boolean indicating whether provided code is free of side-effects
```

kgryte commented on lib/node_modules/@stdlib/repl/lib/eager_evaluator.js:
```diff
@@ -0,0 +1,326 @@
+}
+
+/**
+* Checks if the code is side-effect-free.
```
```suggestion
* Checks whether provided code is free of side-effects.
```

kgryte committed:
Apply suggestions from code review

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/repl/lib/eager_evaluator.js:
```diff
@@ -0,0 +1,325 @@
+		debug( 'Encountered an error when generating AST: %s', err.message );
+		return false;
+	}
+	// Iterate from each node in the body:
```
```suggestion
```

kgryte committed:
docs: remove comment

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/repl/lib/eager_evaluator.js:
```diff
@@ -0,0 +1,324 @@
+/**
+* Disables the eager evaluator.
+*
+* @name disable
```
```suggestion
* @private
* @name disable
```

kgryte commented on lib/node_modules/@stdlib/repl/lib/eager_evaluator.js:
```diff
@@ -0,0 +1,324 @@
+/**
+* Enables the eager evaluator.
+*
+* @name enable
```
```suggestion
* @private
* @name enable
```

kgryte committed:
Apply suggestions from code review

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/repl/lib/eager_evaluator.js:
```diff
@@ -0,0 +1,326 @@
+* @returns {void}
+*/
+setNonEnumerableReadOnly( EagerEvaluator.prototype, 'beforeKeypress', function beforeKeypress() {
+	if (!this._isPreviewing ) {
```
```suggestion
	if ( !this._isPreviewing ) {
```

kgryte committed:
style: fix missing space

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/repl/lib/eager_evaluator.js:
```diff
@@ -0,0 +1,326 @@
+* @returns {void}
+*/
+setNonEnumerableReadOnly( EagerEvaluator.prototype, 'beforeKeypress', function beforeKeypress() {
+	if ( !this._isPreviewing ) {
+		return;
+	}
```
```suggestion
```

This branch isn't necessary given the current logic.

kgryte committed:
refactor: remove unnecessary branch

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/repl/lib/eager_evaluator.js:
```diff
@@ -0,0 +1,323 @@
+	code = cmd.join( '\n' );
+	debug( 'Eagerly evaluating: %s', code );
+	if ( !this._isSideEffectFree( code ) ) {
+		debug( 'Code is not side-effect free, exiting eager-evaluation...' );
```
```suggestion
		debug( 'Unable to perform eager-evaluation due to potential side-effects. Skipping...' );
```

kgryte committed:
refactor: update debug message

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/repl/lib/eager_evaluator.js:
```diff
@@ -0,0 +1,323 @@
+		debug( 'Unable to perform eager-evaluation due to potential side-effects. Skipping...' );
+		return;
+	}
+	debug( 'Trying to process command...' );
+	tmp = processCommand( code );
+	if ( tmp instanceof Error ) {
+		debug( 'Error encountered when processing command: %s', tmp.message );
+		return;
+	}
+	debug( 'Trying to compile command...' );
```
```suggestion
	debug( 'Processing command...' );
	tmp = processCommand( code );
	if ( tmp instanceof Error ) {
		debug( 'Error encountered when processing command: %s', tmp.message );
		return;
	}
	debug( 'Compiling command...' );
```

kgryte committed:
refactor: update debug messages

Signed-off-by: Athan <kgryte@gmail.com>

kgryte reviewed (approved):
LGTM

kgryte commented:
Thanks, @Vinit-Pandit and @Snehil-Shah!
